tt-train/sources/ttml/core/distributed_mapping.hpp

```diff
@@ -0,0 +1,261 @@
+        for (int i = 0; i < rows; ++i) {
+            std::vector<xt::xarray<T>> row_tensors;
+            for (int j = 0; j < cols; ++j) {
+                int index = i * cols + j;
+                row_tensors.push_back(tensors[index]);
+            }
+            mesh_shape_tensors.push_back(row_tensors);
+        }
+
+        // Concatenate along columns first (within each row)
+        std::vector<xt::xarray<T>> row_concatenated;
+        for (const auto& row : mesh_shape_tensors) {
+            auto concatenated_row = xt::concatenate(row, col_dim);
+            row_concatenated.push_back(concatenated_row);
+        }
+
+        // Then concatenate the resulting tensors along rows
+        auto result = xt::concatenate(row_concatenated, row_dim);
+        return result;
+    }
+
+private:
+    std::pair<int, int> m_mesh_shape;
+    std::pair<int, int> m_dims;
+};
+
+// ReplicateTensorToMesh using CRTP
+template <typename T>
+class ReplicateTensorToMesh : public TensorToMesh<ReplicateTensorToMesh<T>, T> {
+public:
+    using Base = TensorToMesh<ReplicateTensorToMesh<T>, T>;
+    ReplicateTensorToMesh(const std::shared_ptr<ttnn::distributed::MeshDevice>& mesh_device) : Base(mesh_device) {
+    }
+
+    std::vector<xt::xarray<T>> map_impl(const xt::xarray<T>& tensor) {
+        int num_devices = Base::m_mesh_device->num_devices();
+        std::vector<xt::xarray<T>> tensors;
+        for (int i = 0; i < num_devices; ++i) {
+            tensors.push_back(tensor);  // Note: this copies the tensor
+        }
+        return tensors;
+    }
+
+    std::unordered_map<std::string, std::string> config_impl() {
+        return {{"strategy", "replicate"}, {"replication_factor", std::to_string(Base::m_mesh_device->num_devices())}};
+    }
+};
+
+template <typename T>
+class ConcatMeshToTensor : public MeshToTensor<ConcatMeshToTensor<T>, T> {
+public:
+    using Base = MeshToTensor<ConcatMeshToTensor<T>, T>;
+    ConcatMeshToTensor(const std::shared_ptr<ttnn::distributed::MeshDevice>& mesh_device, int dim) :
+        Base(mesh_device), m_concat_dim(dim) {
+    }
+
+    xt::xarray<T> compose_impl(const std::vector<xt::xarray<T>>& tensors) {
+        auto result = xt::concatenate(tensors, m_concat_dim);
+        return result;
+    }
+
+private:
+    int m_concat_dim = 0;
+};
+
+template <typename T>
+class VectorMeshToTensor : public MeshToTensor<VectorMeshToTensor<T>, T> {
+public:
+    using Base = MeshToTensor<VectorMeshToTensor<T>, T>;
+    VectorMeshToTensor(const std::shared_ptr<ttnn::distributed::MeshDevice>& mesh_device) : Base(mesh_device) {
+    }
+
+    std::vector<xt::xarray<T>> compose(const std::vector<xt::xarray<T>>& tensors) {
+        return tensors;
+    }
+};
+
+template <typename T>
+using TensorToMeshVariant = std::variant<ShardTensorToMesh<T>, ShardTensor2dMesh<T>, ReplicateTensorToMesh<T>>;
+template <typename T>
+using MeshToTensorVariant = std::variant<ConcatMeshToTensor<T>, ConcatMesh2dToTensor<T>, VectorMeshToTensor<T>>;
```

Comment:
Are you preferring this over traditional inheritance for efficiency?

Reply:
Problem here that type T could be float, uint32_t, int32_t.